File: crystal-ball/src/main/java/org/strangeforest/tcb/model/SetOutcome.java
package org.strangeforest.tcb.model;

public class SetOutcome extends DiffOutcome {

	private final double pServe;
	private final double pReturn;
	private final boolean tieBreak;

	public SetOutcome(double pServe, double pReturn) {
		this(pServe, pReturn, true);
	}

	public SetOutcome(double pServe, double pReturn, boolean tieBreak) {
		super(6, 2, gameNo -> gameNo % 2 == 0 ? new GameOutcome(pServe).pWin() : new GameOutcome(pReturn).pWin());
		this.pServe = pServe;
		this.pReturn = pReturn;
		this.tieBreak = tieBreak;
	}

	@Override protected double pDeuce(double p1, double p2) {
		if (tieBreak)
			return new TieBreakOutcome(pServe, pReturn).pWin();
		else
			return super.pDeuce(p1, p2);
	}
}

File: crystal-ball/src/main/kotlin/org/strangeforest/tcb/model/SetOutcome.kt
package org.strangeforest.tcb.model

class SetOutcome (
	val pServe: Double,
	val pReturn: Double,
	val tieBreakAt: Int? = 6
) : DiffOutcome(6, 2, { gameNo -> if (gameNo % 2 == 0) GameOutcome(pServe).pWin() else GameOutcome(pReturn).pWin() }) {

	override fun pDeuce(p1: Double, p2: Double, items1: Int, items2: Int): Double =
		if (tieBreakAt == items1 && tieBreakAt == items2)
			TieBreakOutcome(pServe, pReturn).pWin()
		else
			super.pDeuce(p1, p2, items1, items2)
}